include/mouse.h

```diff
@@ -136,6 +142,35 @@ bool MOUSEPS2_SendPacket();
 void MOUSEDOS_BeforeNewVideoMode();
 void MOUSEDOS_AfterNewVideoMode(const bool is_mode_changing);
 
+// ***************************************************************************
+// VMM mouse interfaces
+// ***************************************************************************
+
+struct VirtualBoxPointerStatus {
+	uint16_t absolute_x = 0;
+	uint16_t absolute_y = 0;
+};
+
+bool MOUSE_VirtualBox_IsSupported();
+void MOUSE_VirtualBox_ClientDisconnected();
+void MOUSE_VirtualBox_SetPointerAbsolute(const bool is_absolute);
+void MOUSE_VirtualBox_SetPointerVisible(const bool is_visible);
+void MOUSE_VirtualBox_GetPointerStatus(VirtualBoxPointerStatus& status);
+
```

Comment:
normally the module naming is:

`MODULE_CamelCaseFunction`

But I see that `VirtualBox` here is kind of a submodule, given there's also a bunch of `MOUSE_VMware_FunctionNames` below.

So .. ok, I guess this can work.

But then I see further down VirtualBox having it's own `VIRTUALBOX_FunctionNames`:

```c++
VIRTUALBOX_NotifyBooting();
```

So it's getting a bit confusing.. `MOUSE_VirtualBox_` vs `VIRTUALBOX_`, are both public module APIs.

Is it possible to adhere to one or the other?

Reply:
It's like `MOUSEBIOS_*` and `BIOS` - currently we only have one device emulated via _VMware_ and _VirtualBox_ interfaces (in both cases this is a mouse protocol extension), but the idea is that in the future we might want to emulate more devices - see comments in the code below. The idea behind `virtualbox.cpp` and `vmware.cpp` is that they should provide hardware interfaces (I/O ports, PCI devices), decode the guest side requests/commands, and send them to appropriate lower-level modules.

I have no better idea for the naming, but I am open to proposals.

```C++
void VIRTUALBOX_Init(Section* sec)
{
	has_feature_mouse = MOUSE_VirtualBox_IsSupported();
	if (has_feature_mouse) {
		state.mouse_features.SetInitialValue();
	}

	// TODO: implement more features:
	// - shared directories for VBSF.EXE driver by Javis Pedro
	// - possibly display for the OS/2 Museum work-in-progres drivers
	//   https://www.os2museum.com/wp/antique-display-driving/
```

```C++
void VMWARE_Init(Section* sec)
{
	has_feature_mouse = MOUSE_VmWare_IsSupported();

	// TODO: implement more features:
	// - shared directories, for VMSMount tool:
	//   https://github.com/eduardocasino/vmsmount
	// - everything supported by Windows 9x Guest Additions
```

Reply:
Thanks for explain that, @FeralChild64. 

The MIDI system is similar in terms of having multiple possible player "models", so it's a decent comparison just for discussion:

 - Similar to `MOUSE_`, the central `MIDI_` module parses `mididevice = ...` and checks which models are programmatically available using a registry, and then tries activating the desired model and checking its startup state.

 - The player models, like VirtualBox and VmWare here, are each in their own module ((`FLUID_`, `MT32_`, ...). They're derived from a base `MidiHandler` class and registered in a central list that the `MIDI_` model queries via `std::string` name.

    The player models are registered on startup, but they are only "started" by the `MIDI_` if requested (ie: SF2 loaded? / MT32 ROMs loaded?, etc..).

 - The `MPU401_` module opens the IO ports and transmits MIDI messages to the active MIDI player, similar to the `PS/2`, `COM/SERIAL_`, `BIOS_` mouse transports.


The nice aspect of the MIDI design is the N player models all register themselves generically, which avoids having model-specific details in the central `MIDI_` registry.

Specifically, the central `MIDI_` module doesn't have things like `MIDI_ALSA_IsSupported()` that the `ALSA_` model then queries, because that's kind of polluting the central (and generic) `MIDI_` module with ALSA-specific knowledge.

The parallel is here avoiding the central `MOUSE_` module from having stuff like `MOUSE_VirtualBox_IsSupported()` in it and that VirtualBox-specific is-supported logic sitting in the VIRTUALBOX_ module instead.

I'm not saying you need to switch to it though; but just sharing it for thought.   Perhaps this approach would make the mouse area less maintainable and harder to understand (for example, if a bunch of common mouse code had to be duplicated into each vendor-module).

You're in charge of this area though - so maybe the design is actually more maintainable and understandable how you've written it.

Just food for thought, if it's of value! (feel free to resolve if this isn't helpful).

Reply:
Yep, seems like a good candidate for a similar "one interface with multiple implementations" design.

Reply:
I can make `mouse_virtual_machines.cpp` more generic, but making it completely driver agnostic wouldn't be practical - contrary to MIDI devices, VMWare and VirtualBox mouse interfaces provide different capabilities.

Reply:
> but making it completely driver agnostic wouldn't be practical - contrary to MIDI devices, VMWare and VirtualBox mouse interfaces provide different capabilities.

Sounds good @FeralChild64; setting this to resolved - the bottom line is code maintainability. Contorting things just to fit a one-off design pattern certainly isn't worth it, if it'll worsening the code or maintainability! (sorry if this derailed some of your progress, as well).